predicators/approaches/vlm_open_loop_approach.py

```diff
@@ -0,0 +1,192 @@
+
+    NOTE: we don't test this approach because it requires an environment
+    that has rendering, and our current envs that have this are burger
+    and kitchen, which are slow/untestable on GitHub remote...
+    """
+
+    def __init__(self, initial_predicates: Set[Predicate],
+                 initial_options: Set[ParameterizedOption], types: Set[Type],
+                 action_space: Box, train_tasks: List[Task]) -> None:
+        super().__init__(initial_predicates, initial_options, types,
+                         action_space, train_tasks)
+        # Set up the vlm and base prompt.
+        self._vlm = utils.create_vlm_by_name(CFG.vlm_model_name)
+        filepath_to_vlm_prompt = utils.get_path_to_predicators_root() + \
+        "/predicators/approaches/vlm_planning_prompts/no_few_shot.txt"
+        if CFG.vlm_open_loop_use_training_demos:
+            filepath_to_vlm_prompt = utils.get_path_to_predicators_root() + \
+        "/predicators/approaches/vlm_planning_prompts/few_shot.txt"
+        with open(filepath_to_vlm_prompt, "r", encoding="utf-8") as f:
+            self.base_prompt = f.read()
+        self._prompt_state_imgs_list: List[PIL.Image.Image] = []
+        self._prompt_demos_str = ""
+
+    @classmethod
+    def get_name(cls) -> str:
+        return "vlm_open_loop"
+
+    @property
+    def is_learning_based(self) -> bool:
+        return True
+
+    def learn_from_offline_dataset(self, dataset: Dataset) -> None:
+        """Adds the images and plans from the training dataset to the base
+        prompt for use at test time!"""
+        if not CFG.vlm_open_loop_use_training_demos:
+            return None
+        # Crawl thru the dataset and pull out all the images.
+        # For each image, add text to it in the bototm left indicating the
+        # trajectory and timestep it's from.
+        assert dataset.trajectories[0].states[0].simulator_state is not None
+        assert isinstance(
+            dataset.trajectories[0].states[0].simulator_state["images"], List)
+        num_imgs_per_state = len(
+            dataset.trajectories[0].states[0].simulator_state["images"])
+        self._prompt_demos_str = ""
+        for traj_num, traj in enumerate(dataset.trajectories):
+            traj_goal = self._train_tasks[traj.train_task_idx].goal
+            self._prompt_demos_str += f"Demonstration {traj_num}, " + \
+                f"Goal: {str(sorted(traj_goal))}\n"
+            for state_num, state in enumerate(traj.states):
+                assert state.simulator_state is not None
+                assert len(
+                    state.simulator_state["images"]) == num_imgs_per_state
+                for img_num, img in enumerate(state.simulator_state["images"]):
+                    pil_img = PIL.Image.fromarray(img)  # type: ignore
+                    draw = ImageDraw.Draw(pil_img)
+                    img_font = utils.get_scaled_default_font(draw, 20)
+                    img_with_txt = utils.add_text_to_draw_img(
```

Comment:
We probably don't want to draw onto the image itself, in case the text overlaps with objects and/or labels for objects (e.g. our set of marks labels)?

It would be better to add a white rectangle to the top or bottom of the image and put this into into that white rectangle.

Reply:
yes - good idea!